client/modules/IDE/components/Editor/index.jsx

```diff
@@ -469,152 +258,371 @@ class Editor extends React.Component {
+    props.provideController({
+      tidyCode,
+      showFind,
+      showReplace,
+      getContent
     });
 
-    const { currentLine } = this.state;
-
-    return (
-      <MediaQuery minWidth={770}>
-        {(matches) =>
-          matches ? (
-            <section className={editorSectionClass}>
-              <div className="editor__header">
-                <button
-                  aria-label={this.props.t('Editor.OpenSketchARIA')}
-                  className="sidebar__contract"
-                  onClick={() => {
-                    this.props.collapseSidebar();
-                    this.props.closeProjectOptions();
-                  }}
-                >
-                  <LeftArrowIcon focusable="false" aria-hidden="true" />
-                </button>
-                <button
-                  aria-label={this.props.t('Editor.CloseSketchARIA')}
-                  className="sidebar__expand"
-                  onClick={this.props.expandSidebar}
-                >
-                  <RightArrowIcon focusable="false" aria-hidden="true" />
-                </button>
-                <div className="editor__file-name">
-                  <span>
-                    {this.props.file.name}
-                    <UnsavedChangesIndicator />
-                  </span>
-                  <Timer />
-                </div>
-              </div>
-              <article
-                ref={(element) => {
-                  this.codemirrorContainer = element;
+    return () => {
+      if (cmRef.current) {
+        cmRef.current.off('keyup', handleKeyUp);
+      }
+      props.provideController(null);
+    };
+  }, []);
+
+  useEffect(() => {
+    if (cmRef.current) {
+      cmRef.current.on('change', handleChange);
+    }
+    return () => {
+      cmRef.current.off('change', handleChange);
+    };
+  }, [props.autorefresh, props.isPlaying, props.file.id]);
+
+  useEffect(() => {
+    if (cmRef.current) {
+      cmRef.current.on('keydown', handleKeydown);
+    }
+    return () => {
+      cmRef.current.off('keydown', handleKeydown);
+    };
+  }, [props.autocompleteHinter]);
+
+  useEffect(() => {
+    initializeDocuments(props.files);
+  }, [props.files]);
+
+  useEffect(() => {
+    if (!cmRef.current) return;
+
+    const fileMode = getFileMode(props.file.name);
+    if (fileMode === 'javascript') {
+      cmRef.current.setOption('emmet', {
+        preview: ['html'],
+        markTagPairs: false,
+        autoRenameTags: true
+      });
+    }
+
+    if (props.file.id !== prevFileId.current) {
+      const oldDoc = cmRef.current.swapDoc(docsRef.current[props.file.id]);
+      docsRef.current[prevFileId.current] = oldDoc;
+      cmRef.current.focus();
+
+      if (!props.unsavedChanges) {
+        setTimeout(() => props.setUnsavedChanges(false), 400);
+      }
+
+      prevFileId.current = props.file.id;
+      cmRef.current.focus();
+    }
+  }, [props.file.id, props.file.name, props.unsavedChanges]);
+
+  // useEffect(() => {
```

Comment:
To maintain readability, we'd ideally only want to include comments that provide supportive information about the code!